docs/chinese/config.md


## 配置项说明
```yaml
# 网络监听器
listeners:
  # (可选) 监听 TCP 地址
  mqtt:
    # 绑定的 Socket 地址
    addr: 127.0.0.1:1883
    # 允许多个 listener 绑定在同一个端口上，可以增加建立链接的速度
    reuse_port: false
    # (可选) proxy protocol 模式, 可能的选项:
    #    null           : 不启用 proxy protocol
    #    Normal         : 客户端非 TLS, 服务端非 TLS
    #    TlsTermination : 客户端 TLS, proxy 处理 TLS, 服务端非 TLS (会从 proxy protocol header 中读取 host name(SNI))
    proxy_mode: null
  # (可选) 监听 TCP+TLS 地址
  mqtts:
    # 绑定的 Socket 地址
    addr: 127.0.0.1:8883
    # 允许多个 listener 绑定在同一个端口上，可以增加建立链接的速度
    reuse_port: false
    # 是否开启 proxy protocol v2
    proxy: false
    # 用来认证客户端的 CA 文件, 如果 `verify_peer` 是 true 这个字段必须填上
    ca_file: /path/to/mqtts.ca
    # key 文件
    key_file: /path/to/mqtts.key
    # cert 文件
    cert_file: /path/to/mqtts.cert
    # 校验客户端的的证书是被信任的
    verify_peer: true
    # 如果在握手阶段客户端没发送它的证书马上终止连接。需要先开启 `verify_peer` 这个配置项才有效.
    fail_if_no_peer_cert: true
  # (同 `listeners.mqtt`) WebSocket 监听器
  ws: null
  # (同 `listeners.mqtts`) WebSocket+TLS 监听器
  wss: null
  # (可选) API入口点
  http:
    # 绑定的 Socket 地址
    addr: 127.0.0.1:8881
    # 允许多个 listener 绑定在同一个端口上，可以增加建立链接的速度
    reuse_port: false
    # 启用 Prometheus 监控，在 /metrics 路径下
    prometheus: true
# 基于密码文件的认证，密码用来校验 connect 数据包中的 username/password 字段
auth:
  enable: true
  # 密码文件, 请使用 insert-password/remove-password 子命令来管理密码
  password_file: /path/to/passwords/file

# (v5.0 专有) 用于 MQTT v5.0 增强认证的 Scram 用户列表
#   生成 hashed password 的方法: https://github.com/akasamq/akasa/blob/8503adb566c46074d57bfea8dbe39a6fd3403e28/akasa-core/src/tests/protocols/mqtt/v5/v500/auth.rs#L21-L24
scram_users:
  user:
    hashed_password: 2a2a2a
    iterations: 4096
    salt: 73616c74
# (v5.0 专有) MQTT v5.0 增强认证支持的 sasl 机制, 当前只支持 `SCRAM-SHA-256`
sasl_mechanisms:
- SCRAM-SHA-256

# 通过 MQTT v3.1 协议连接的时候, 如果设置这个选项为 true 服务器会拒绝所有 client identifier 长度超过 23 字节的连接.
check_v310_client_id_length: false
# (v5.0 专有) 共享订阅模式, 可选项: [Random, HashClientId, HashTopicName]
shared_subscription_mode: Random
# 客户端允许使用的最高 QoS 级别
max_allowed_qos: 2
# 重发消息的超时时间 (单位: 秒)
inflight_timeout: 15
# 最大允许服务端同时发送给客户端的消息数量的默认值
max_inflight_client: 10
# 最大允许客户端同时发送给服务端的消息数量
max_inflight_server: 10
# 最大允许的存储在内存中的待发消息
max_in_mem_pending_messages: 256
# (未使用) 最大允许的存储在数据库中的待发消息
max_in_db_pending_messages: 65536
# (v5.0 专有) 最小允许的 keep alive 值
min_keep_alive: 10
# (v5.0 专有) 最大允许的 keep alive 值
max_keep_alive: 65535
# (v5.0 专有, 未使用)
multiple_subscription_id_in_publish: false
# (v5.0 专有) 可以在 connect packet 中设置的最大的 session expiry interval 值 (单位: 秒)
max_session_expiry_interval: 4294967295
# (v5.0 专有) 客户端限制服务端可以发送的最大 packet 提价 (单位: 字节)
max_packet_size_client: 268435460
# (v5.0 专有) 服务端限制客户端可以发送的最大 packet 体积 (单位: 字节)
max_packet_size_server: 268435460
# (v5.0 专有) publish 消息中 topic alias 的最大值
topic_alias_max: 65535
# (v5.0 专有) 是否支持保留消息
retain_available: true
# (v5.0 专有) 是否支持共享订阅
shared_subscription_available: true
# (v5.0 专有) 是否支持订阅 ID
subscription_id_available: true
# (v5.0 专有) 是否支持通配符订阅
wildcard_subscription_available: true
# 控制哪些 hook 函数被调用
hook:
  enable_before_connect: true
  enable_after_connect: true
  enable_publish: true
  enable_subscribe: true
  enable_unsubscribe: true
```


docs/english/config.md


## Config Options Explanation
```yaml
# Network Listeners
listeners:
  # (optional) Listen on TCP socket
  mqtt:
    # The socket address to bind
    addr: 127.0.0.1:1883
    # Allows the socket to bind to an in-use port to increase connection accept speed
    reuse_port: false
    # (optional) proxy protocol mode, can be:
    #    null           : Disable proxy protocol
    #    Normal         : Client side non-TLS, server side non-TLS
    #    TlsTermination : Client side TLS, proxy handle TLS, server side non-TLS (read host name(SNI) from proxy protocol header)
    proxy_mode: null
  # (optional) Listen on TCP socket with TLS
  mqtts:
    # The socket address to bind
    addr: 127.0.0.1:8883
    # Allows the socket to bind to an in-use port to increase connection accept speed
    reuse_port: false
    # Enable proxy protocol v2 or not
    proxy: false
    # This CA file is for verify client certificate, if `verify_peer` is true this field MUST be presented.
    ca_file: /path/to/mqtts.ca
    # key file
    key_file: /path/to/mqtts.key
    # cert file
    cert_file: /path/to/mqtts.cert
    # Verifies that the client’s certificate is trusted.
    verify_peer: true
    # Abort the handshake if the client did not send a certificate. This should be paired with `verify_peer`.
    fail_if_no_peer_cert: true
  # (same with `listeners.mqtt`) WebSocket listener
  ws: null
  # (same with `listeners.mqtts`) WebSocket with TLS listener
  wss: null
  # (optional) API entry point
  http:
    # The socket address to bind
    addr: 127.0.0.1:8881
    # Allows the socket to bind to an in-use port to increase connection accept speed
    reuse_port: false
    # Enable prometheus on location /metrics
    prometheus: true
# Password file based authentication, the config used to check username/password fields in connect packet.
auth:
  enable: true
  # The password file, please use insert-password/remove-password subcommand to manage the passwords
  password_file: /path/to/passwords/file

# (v5.0 only) Scram users used in MQTT v5.0 enhanced authentication.
#   To generate the hashed password: https://github.com/akasamq/akasa/blob/8503adb566c46074d57bfea8dbe39a6fd3403e28/akasa-core/src/tests/protocols/mqtt/v5/v500/auth.rs#L21-L24
scram_users:
  user:
    hashed_password: 2a2a2a
    iterations: 4096
    salt: 73616c74
# (v5.0 only) The sasl mechanisms supported in MQTT v5.0 enhanced authentication, now only `SCRAM-SHA-256` is supported
sasl_mechanisms:
- SCRAM-SHA-256

# When client connect with MQTT v3.1 protocol, if set this option to true, server will forbid client identifier length greater than 23.
check_v310_client_id_length: false
# (v5.0 only) The shared subscription mode, can be: [Random, HashClientId, HashTopicName]
shared_subscription_mode: Random
# Maximum allowed QoS the client can publish or subscribe
max_allowed_qos: 2
# Timeout seconds to resend inflight pending messages (unit: second)
inflight_timeout: 15
# Maximum inflight pending messages for client default value
max_inflight_client: 10
# Maximum inflight pending messages the server will handle
max_inflight_server: 10
# Maximum allowed pending messages in memory
max_in_mem_pending_messages: 256
# (unused) Maximum allowed pending messages in database
max_in_db_pending_messages: 65536
# (v5.0 only) The minimum allowed keep alive
min_keep_alive: 10
# (v5.0 only) The maximum allowed keep alive
max_keep_alive: 65535
# (v5.0 only, unused)
multiple_subscription_id_in_publish: false
# (v5.0 only) The maximum session expiry interval value can set in connect packet (unit: second)
max_session_expiry_interval: 4294967295
# (v5.0 only) The maximum packet size given by client (to limit server, unit: byte)
max_packet_size_client: 268435460
# (v5.0 only) The maximum packet size given by server (to limit client, unit: byte)
max_packet_size_server: 268435460
# (v5.0 only) The maximum topic alias value can be used in publish packet
topic_alias_max: 65535
# (v5.0 only) Whether support retained message
retain_available: true
# (v5.0 only) Whether support shared subscription
shared_subscription_available: true
# (v5.0 only) Whether support subscription identifiers
subscription_id_available: true
# (v5.0 only) Whether supports wildcard subscriptions
wildcard_subscription_available: true
# The value indicate whether call certain hook function
hook:
  enable_before_connect: true
  enable_after_connect: true
  enable_publish: true
  enable_subscribe: true
  enable_unsubscribe: true
```


akasa-core/src/config.rs
use std::collections::{HashMap, HashSet};
use std::net::{Ipv4Addr, SocketAddr};
use std::path::PathBuf;
use std::str::FromStr;

use mqtt_proto::QoS;
use serde::{Deserialize, Serialize};

pub const DEFAULT_MAX_PACKET_SIZE: u32 = 5 + 268_435_455;

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Config {
    pub listeners: Listeners,
    pub auth: AuthConfig,
    pub scram_users: HashMap<String, ScramPasswordInfo>,
    pub sasl_mechanisms: HashSet<SaslMechanism>,
    /// It seems all populte MQTT server(broker) not check this.
    pub check_v310_client_id_length: bool,

    pub shared_subscription_mode: SharedSubscriptionMode,

    /// max allowed qos, allowed values: [0, 1, 2], default: 2
    pub max_allowed_qos: u8,

    /// Timeout seconds to re-send inflight pending messages
    pub inflight_timeout: u64,
    /// max inflight pending messages for client default value
    pub max_inflight_client: u16,
    /// max inflight pending messages the server will handle
    pub max_inflight_server: u16,
    /// max allowed pending messages in memory, default: 256
    pub max_in_mem_pending_messages: usize,
    /// max allowed pending messages in database, default: 65536
    pub max_in_db_pending_messages: usize,

    pub min_keep_alive: u16,
    pub max_keep_alive: u16,
    pub multiple_subscription_id_in_publish: bool,

    pub max_session_expiry_interval: u32,
    /// max packet size given by client (to limit server)
    pub max_packet_size_client: u32,
    /// max packet size given by server (to limit client)
    pub max_packet_size_server: u32,
    pub topic_alias_max: u16,
    pub retain_available: bool,
    pub shared_subscription_available: bool,
    pub subscription_id_available: bool,
    pub wildcard_subscription_available: bool,

    pub hook: HookConfig,
}

#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq)]
pub struct Listeners {
    /// default port: 1883
    pub mqtt: Option<Listener>,
    /// default port: 8883
    pub mqtts: Option<TlsListener>,
    /// default port: 8080
    pub ws: Option<Listener>,
    /// default port: 8443
    pub wss: Option<TlsListener>,
    /// default port: 8081
    pub http: Option<Http>,
}

#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq)]
pub struct Listener {
    pub addr: SocketAddr,
    pub reuse_port: bool,
    /// The proxy protocol v2 mode
    pub proxy_mode: Option<ProxyMode>,
}

#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq)]
pub struct TlsListener {
    pub addr: SocketAddr,
    pub reuse_port: bool,
    /// Enable proxy protocol v2 or not
    pub proxy: bool,
    /// This CA file is for verify client certificate, if `verify_peer` is true
    /// this field MUST be presented.
    pub ca_file: Option<PathBuf>,
    pub key_file: PathBuf,
    pub cert_file: PathBuf,
    pub verify_peer: bool,
    pub fail_if_no_peer_cert: bool,
}

#[derive(Serialize, Deserialize, Clone, Copy, Debug, Eq, PartialEq)]
pub enum ProxyMode {
    /// Client side non-TLS, server side non-TLS
    Normal,
    /// Client side TLS, proxy handle TLS, server side non-TLS (read host name(SNI) from proxy protocol header)
    TlsTermination,
}

#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq)]
pub struct Http {
    pub addr: SocketAddr,
    pub reuse_port: bool,
    pub prometheus: bool,
}

impl Default for Http {
    fn default() -> Self {
        Self {
            addr: "0.0.0.0:8081".parse().unwrap(),
            reuse_port: true,
            prometheus: true,
        }
    }
}

impl Default for Listeners {
    fn default() -> Listeners {
        Listeners {
            mqtt: Some(Listener {
                addr: (Ipv4Addr::LOCALHOST, 1883).into(),
                proxy_mode: None,
                reuse_port: true,
            }),
            mqtts: None,
            ws: None,
            wss: None,
            http: None,
        }
    }
}

#[derive(Serialize, Deserialize, Clone, Copy, Debug, Eq, PartialEq, Hash)]
pub enum SaslMechanism {
    #[serde(rename = "SCRAM-SHA-256")]
    ScramSha256,

    // TODO: Not supported yet
    #[serde(rename = "SCRAM-SHA-256-PLUS")]
    ScramSha256Plus,
}

#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq)]
pub struct ScramPasswordInfo {
    #[serde(with = "hex::serde")]
    pub hashed_password: Vec<u8>,

    pub iterations: u16,

    #[serde(with = "hex::serde")]
    pub salt: Vec<u8>,
}

#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq)]
pub struct AuthConfig {
    pub enable: bool,
    pub password_file: Option<PathBuf>,
}

#[derive(Serialize, Deserialize, Clone, Copy, Debug, Eq, PartialEq)]
pub enum SharedSubscriptionMode {
    Random,
    HashClientId,
    HashTopicName,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HookConfig {
    pub enable_before_connect: bool,
    pub enable_after_connect: bool,
    pub enable_after_disconnect: bool,
    pub enable_publish: bool,
    pub enable_subscribe: bool,
    pub enable_unsubscribe: bool,
}

impl Default for HookConfig {
    fn default() -> HookConfig {
        HookConfig {
            enable_before_connect: true,
            enable_after_connect: true,
            enable_after_disconnect: true,
            enable_publish: true,
            enable_subscribe: true,
            enable_unsubscribe: true,
        }
    }
}

impl Default for Config {
    fn default() -> Config {
        let config = Config {
            listeners: Listeners::default(),
            auth: AuthConfig {
                enable: true,
                password_file: Some(PathBuf::from("/path/to/passwords/file")),
            },
            scram_users: vec![("user", (b"***", 4096, b"salt"))]
                .into_iter()
                .map(|(u, (p, i, s))| {
                    (
                        u.to_owned(),
                        ScramPasswordInfo {
                            hashed_password: p.to_vec(),
                            iterations: i,
                            salt: s.to_vec(),
                        },
                    )
                })
                .collect(),
            sasl_mechanisms: vec![SaslMechanism::ScramSha256].into_iter().collect(),
            shared_subscription_mode: SharedSubscriptionMode::Random,
            check_v310_client_id_length: false,
            max_allowed_qos: 2,
            inflight_timeout: 15,
            max_inflight_client: 10,
            max_inflight_server: 10,
            max_in_mem_pending_messages: 256,
            max_in_db_pending_messages: 65536,
            min_keep_alive: 10,
            max_keep_alive: u16::MAX,
            multiple_subscription_id_in_publish: false,
            max_session_expiry_interval: u32::MAX,
            max_packet_size_client: DEFAULT_MAX_PACKET_SIZE,
            max_packet_size_server: DEFAULT_MAX_PACKET_SIZE,
            topic_alias_max: u16::MAX,
            retain_available: true,
            shared_subscription_available: true,
            subscription_id_available: true,
            wildcard_subscription_available: true,

            hook: HookConfig::default(),
        };
        assert!(config.is_valid(), "default config");
        config
    }
}

impl Config {
    pub fn new_allow_anonymous() -> Config {
        Config {
            auth: AuthConfig {
                enable: false,
                password_file: None,
            },
            ..Default::default()
        }
    }

    /// Check if the config is valid
    pub fn is_valid(&self) -> bool {
        if self.auth.enable && self.auth.password_file.is_none() {
            log::error!("when authentication enabled, `password_file` must be provided");
            return false;
        }
        if self.max_allowed_qos > 2 {
            log::error!(
                "invalid max_allowed_qos: {}, allowed values: [0, 1, 2]",
                self.max_allowed_qos
            );
            return false;
        }
        if self.max_packet_size_client == 0 {
            log::error!("invalid client max_packet_size, 0 is not allowed");
            return false;
        }
        if self.max_packet_size_server == 0 {
            log::error!("invalid server max_packet_size, 0 is not allowed");
            return false;
        }
        for mechanism in &self.sasl_mechanisms {
            if mechanism != &SaslMechanism::ScramSha256 {
                log::error!("invalid sasl_mechanism, only `SCRAM-SHA-256` is allowed");
                return false;
            }
        }
        for password_info in self.scram_users.values() {
            if password_info.iterations < 4096 {
                // RFC-7677: For the SCRAM-SHA-256 and SCRAM-SHA-256-PLUS SASL
                // mechanisms, the hash iteration-count announced by a server
                // SHOULD be at least 4096.
                log::error!("scram_users password iterations must >= 4096 (see RFC-7677)");
                return false;
            }
        }
        if let Listeners {
            mqtt: None,
            mqtts: None,
            ws: None,
            wss: None,
            http: None,
        } = self.listeners
        {
            log::error!("No listen address found");
            return false;
        }
        true
    }

    pub fn max_allowed_qos(&self) -> QoS {
        match self.max_allowed_qos {
            0 => QoS::Level0,
            1 => QoS::Level1,
            2 => QoS::Level2,
            value => panic!("invalid Config.max_allowed_qos: {value}"),
        }
    }
}

impl FromStr for SaslMechanism {
    type Err = ();

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        match value {
            "SCRAM-SHA-256" => Ok(Self::ScramSha256),
            "SCRAM-SHA-256-PLUS" => Ok(Self::ScramSha256Plus),
            _ => Err(()),
        }
    }
}
